src/write_gcm.c

```diff
@@ -2275,59 +2275,136 @@ static monitored_resource_t *monitored_resource_create_from_fields(
   // Changes va_list into label_t array
   va_start(ap, project_id);
   for (int i = 0; i < num_labels; i++) {
-    labels[i].key = va_arg(ap, const char*);
-    labels[i].value = va_arg(ap, const char*);
+    labels[i].key = va_arg(ap, char*);
+    labels[i].value = va_arg(ap, char*);
   }
   va_end(ap);
   return monitored_resource_create_from_array(type, project_id, labels,
                                               num_labels);
 }
 
+typedef struct {
+  _Bool processing_labels;
+  char *label_name;
+  char *type;
+  llist_t *label_list;
+} wg_mr_parse_context_t;
+
+static int wg_mr_handle_string(void *arg, const unsigned char *val,
+    wg_yajl_callback_size_t length) {
+  wg_mr_parse_context_t *ctx = (wg_mr_parse_context_t*) arg;
+  const char *str = (const char *) val;
+
+  if (ctx->processing_labels) {
+    llist_append(ctx->label_list, llentry_create(ctx->label_name, 
+                                                 strndup(str, length)));
+  } else {
+    // The only other key we expect when not processing 'labels' is 'type'.
+    ctx->type = strndup(str, length);
+  }
+
+  return 1;
+}
+
+static int wg_mr_handle_map_key(void *arg, const unsigned char *data,
+    wg_yajl_callback_size_t length) {
+  wg_mr_parse_context_t *ctx = (wg_mr_parse_context_t *) arg;
+  const char *str = (const char *) data;
+
+  if (ctx->processing_labels) {
+    ctx->label_name = strndup(str, length);
+  } else if (strncmp(str, "labels", length) == 0) {
+    ctx->processing_labels = 1;
+    ctx->label_list = llist_create();
+  } else if (strncmp(str, "type", length) != 0) {
```

Comment:
@igorpeshansky and I had spoken about this offline during the initial writing and we decided that it would be better to be extremely specific about what we're expecting here as keys to provide the client with immediate feedback should we receive an unexpected key.